src/Adapter/Product/CommandHandler/AddProductImageHandler.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product\CommandHandler;
+
+use PrestaShop\PrestaShop\Adapter\Image\Uploader\ProductImageUploader;
+use PrestaShop\PrestaShop\Adapter\Product\Repository\ProductImageRepository;
+use PrestaShop\PrestaShop\Core\Domain\Product\Command\AddProductImageCommand;
+use PrestaShop\PrestaShop\Core\Domain\Product\CommandHandler\AddProductImageHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\Image\ValueObject\ImageId;
+
+/**
+ * Handles @see AddProductImageCommand
+ */
+final class AddProductImageHandler implements AddProductImageHandlerInterface
+{
+    /**
+     * @var ProductImageUploader
+     */
+    private $productImageUploader;
+
+    /**
+     * @var ProductImageRepository
+     */
+    private $productImageRepository;
+
+    /**
+     * @var array
+     */
+    private $contextShopIds;
+
+    /**
+     * @param ProductImageUploader $productImageUploader
+     * @param ProductImageRepository $productImageRepository
+     * @param array $contextShopIds
+     */
+    public function __construct(
+        ProductImageUploader $productImageUploader,
+        ProductImageRepository $productImageRepository,
+        array $contextShopIds
+    ) {
+        $this->productImageUploader = $productImageUploader;
+        $this->productImageRepository = $productImageRepository;
+        $this->contextShopIds = $contextShopIds;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(AddProductImageCommand $command): ImageId
+    {
+        $image = $this->productImageRepository->create($command->getProductId(), $this->contextShopIds);
+
+        //@todo: db transaction rollback if upload fails?
```

Comment:
It would be an improvement but unless this is already possible with ObjectModel (so we can simply turn it on), we cannot do this in the scope of this PR.

If it's not something PS ObjectModel can do today, to implement db transaction rollback means
- only MyISAM-compliant
- possibly mess up with other persistence operations

Too big to be done here 😄 

I guess it's not possible to 1st upload the image, 2nd create the entity ?

Reply:
> I guess it's not possible to 1st upload the image, 2nd create the entity ?

nope. Uploading image requires Image id.